### PR TITLE
Semi-Monthly Feature (FINERACT-1322)

### PR DESCRIPTION
**Pull request by Cadreia**
## Description

* Implements the semi-monthly loan scheduler feature.
* Details of this issue can be found on this [Apache Fineract JIRA ticket](https://issues.apache.org/jira/browse/FINERACT-1322).
* An [article](http://research.muellners.org/implementing-a-bi-monthly-loan-product/) has been written on this.
* Corresponding UI change is [here](https://github.com/openMF/community-app/pull/3422).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/api-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java [hunk 1]
```diff
@@ -334,10 +334,10 @@ public CommandProcessingResult submitApplication(final JsonCommand command) {
                 updateProductRelatedDetails(productRelatedDetail, newLoanApplication);
             }
 
-            this.fromApiJsonDeserializer.validateLoanTermAndRepaidEveryValues(newLoanApplication.getTermFrequency(),
-                    newLoanApplication.getTermPeriodFrequencyType(), productRelatedDetail.getNumberOfRepayments(),
-                    productRelatedDetail.getRepayEvery(), productRelatedDetail.getRepaymentPeriodFrequencyType().getValue(),
-                    newLoanApplication);
+            // this.fromApiJsonDeserializer.validateLoanTermAndRepaidEveryValues(newLoanApplication.getTermFrequency(),
+            // newLoanApplication.getTermPeriodFrequencyType(), productRelatedDetail.getNumberOfRepayments(),
```
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java:338]: Why is this validation commented out?

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java [hunk 2]
```diff
@@ -211,23 +211,24 @@ private void validateRepaymentScheduleWithEMI(ArrayList<HashMap> periods) {
     private void assertRepaymentScheduleValuesWithEMI(HashMap period, LoanDisbursementTestBuilder expectedRepaymentSchedule, int position) {
 
         assertEquals(period.get("dueDate").toString(), expectedRepaymentSchedule.getDueDate());
-        assertEquals(period.get("principalLoanBalanceOutstanding"), expectedRepaymentSchedule.getPrincipalLoanBalanceOutstanding());
+        assertEquals(Float.parseFloat(period.get("principalLoanBalanceOutstanding").toString()),
+                expectedRepaymentSchedule.getPrincipalLoanBalanceOutstanding(), 0.1f);
         LOG.info("{}", period.get("totalOriginalDueForPeriod").toString());
         assertEquals(Float.parseFloat(period.get("totalOriginalDueForPeriod").toString()),
-                expectedRepaymentSchedule.getTotalOriginalDueForPeriod().floatValue(), 0.0f);
+                expectedRepaymentSchedule.getTotalOriginalDueForPeriod().floatValue(), 0.1f);
```
Comment (ptuomola) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java:219]: Why have these test results changed? I would not expect the existing integration tests to have to change when adding new products...

**Review by ptuomola — CHANGES_REQUESTED** [review]
Also have we tested what happens if one of the semi dates falls on a date which does not exist in every month - i.e. 31st? Does the logic work sensibly in that case as well?

**Comment by BLasan** [comment]
@Cadreia Please resolve merge conflicts and requested changes by @ptuomola

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanAccountData.java [hunk 3]
```diff
@@ -1021,7 +1021,7 @@ public static LoanAccountData loanProductWithTemplateDefaults(final LoanProductD
         final Collection<CalendarData> calendarOptions = null;
         final Collection<StaffData> loanOfficerOptions = null;
 
-        final EnumOptionData termPeriodFrequencyType = product.getRepaymentFrequencyType();
```
Comment (BLasan) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanAccountData.java:1024]: Any reassignments?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/data/LoanProductData.java [hunk 4]
```diff
@@ -624,7 +625,8 @@ public LoanProductData(final Long id, final String name, final String shortName,
             boolean isFloatingInterestRateCalculationAllowed, final boolean isVariableInstallmentsAllowed,
             final Integer minimumGapBetweenInstallments, final Integer maximumGapBetweenInstallments,
             final boolean syncExpectedWithDisbursementDate, final boolean canUseForTopup, final boolean isEqualAmortization,
-            Collection<RateData> rateOptions, Collection<RateData> rates, final boolean isRatesEnabled) {
+            Collection<RateData> rateOptions, Collection<RateData> rates, final boolean isRatesEnabled, final LocalDate firstDateForSemi,
```
Comment (BLasan) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/data/LoanProductData.java:628]: Why `rateOptions` and `rates` are not final? Any reassignment?

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java [hunk 5]
```diff
@@ -211,23 +211,24 @@ private void validateRepaymentScheduleWithEMI(ArrayList<HashMap> periods) {
     private void assertRepaymentScheduleValuesWithEMI(HashMap period, LoanDisbursementTestBuilder expectedRepaymentSchedule, int position) {
 
         assertEquals(period.get("dueDate").toString(), expectedRepaymentSchedule.getDueDate());
-        assertEquals(period.get("principalLoanBalanceOutstanding"), expectedRepaymentSchedule.getPrincipalLoanBalanceOutstanding());
+        assertEquals(Float.parseFloat(period.get("principalLoanBalanceOutstanding").toString()),
```
Comment (BLasan) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java:214]: Why converted to a Float comparison?

**Review by BLasan — CHANGES_REQUESTED** [review]

**Comment by awasum** [comment]
@Cadreia , Please resolve merge conflicts here and address review comments above.

**Comment by BLasan** [comment]
@Cadreia Could you please address the review comments?

**Comment by awasum** [comment]
@Cadreia Please fix review comments and merge conflicts so this can be reviewed and merged. Important feature.

**Pull request closed without merging**